lint/src/main/java/permissions/dispatcher/CallNeedsPermissionDetector.java

```diff
@@ -1,54 +1,80 @@
 import com.android.tools.lint.detector.api.JavaContext;
 import com.android.tools.lint.detector.api.Scope;
 import com.android.tools.lint.detector.api.Severity;
+import com.intellij.psi.JavaElementVisitor;
+import com.intellij.psi.PsiAnnotation;
+import com.intellij.psi.PsiClass;
+import com.intellij.psi.PsiElement;
+import com.intellij.psi.PsiMethod;
+import com.intellij.psi.PsiMethodCallExpression;
+import com.intellij.psi.PsiReferenceExpression;
 
 import java.util.ArrayList;
+import java.util.Collections;
 import java.util.EnumSet;
 import java.util.HashSet;
 import java.util.List;
 import java.util.Set;
 
-import lombok.ast.Annotation;
-import lombok.ast.AstVisitor;
-import lombok.ast.ClassDeclaration;
-import lombok.ast.ForwardingAstVisitor;
-import lombok.ast.MethodInvocation;
 
+public class CallNeedsPermissionDetector extends Detector implements Detector.JavaPsiScanner {
 
-public class CallNeedsPermissionDetector extends Detector implements Detector.JavaScanner {
     public static final Issue ISSUE = Issue.create("CallNeedsPermission",
             "Call the corresponding \"withCheck\" method of the generated PermissionsDispatcher class instead",
             "Directly invoking a method annotated with @NeedsPermission may lead to misleading behaviour on devices running Marshmallow and up. Therefore, it is advised to use the generated PermissionsDispatcher class instead, which provides a \"withCheck\" method that safely handles runtime permissions.",
             Category.CORRECTNESS,
             7,
             Severity.ERROR,
-            new Implementation(CallNeedsPermissionDetector.class, EnumSet.of(Scope.ALL_JAVA_FILES)));
+            new Implementation(CallNeedsPermissionDetector.class,
+                    EnumSet.of(Scope.ALL_JAVA_FILES)));
 
     static List<String> generatedClassNames = new ArrayList<String>();
 
+    static List<String> METHODS = Collections.emptyList();
+
+    @Override
+    public List<Class<? extends PsiElement>> getApplicablePsiTypes() {
+        return Collections.<Class<? extends PsiElement>>singletonList(PsiClass.class);
+    }
+
+    public CallNeedsPermissionDetector() {
+
+    }
+
     @Override
-    public AstVisitor createJavaVisitor(JavaContext context) {
+    public JavaElementVisitor createPsiVisitor(final JavaContext context) {
         if (context.getPhase() == 1) {
             // find out which class has RuntimePermissions
             return new AnnotationChecker(context);
-        } else if (context.getPhase() == 2) {
-            // find out which class call method with NeedPermission
-            // exclude class with above with name XxxPermissionsDispatcher in the same package.
-            return new MethodCallChecker(context);
         }
         return null;
     }
 
-    private static class AnnotationChecker extends ForwardingAstVisitor {
+    @Override
+    public List<String> getApplicableMethodNames() {
+        return METHODS;
```

Comment:
if you never plan to add explicit method names here, it's safe to omit this method